provider/adapters/src/fork/by_key.rs

```diff
@@ -162,6 +162,21 @@ where
     }
 }
 
+// TODO(#1893): Instead of going through DynProvider, this should only implement ResourceProvider
+// for the ResourceProviders that P0 and P1 implement. I haven't figured out a way to do this though
+// as we'll need two blanket implementations that could conflict.
+#[cfg(feature = "datagen")] // this is not production-ready
+impl<M, P0, P1> ResourceProvider<M> for ForkByKeyProvider<P0, P1>
```

Comment:
Nit: impls on `ForkByKeyProvider` should also be made on `MultiForkByKeyProvider`